pkg/agent/core/wss.go

```diff
@@ -0,0 +1,92 @@
+package core
+
+// Working set size estimation
+// For more, see: https://www.notion.so/neondatabase/874ef1cc942a4e6592434dbe9e609350
+
+import (
+	"fmt"
+)
+
+type WssEstimatorConfig struct {
+	// MaxAllowedIncreaseFactor is the maximum tolerable increase in slope between windows.
+	// If the slope increases by more than this factor, we will cut off the working set size as the
+	// border between the two windows.
+	MaxAllowedIncreaseFactor float64
+	// InitialOffset is the index of the minimum working set size we must consider.
+	//
+	// In practice, this is taken from the scaling config's LFCMinWaitBeforeDownscaleMinutes, with
+	// the expectation that datapoints are all one minute apart, starting at 1m. So a value of 15m
+	// translates to an InitialOffset of 14 (-1 because indexes start at zero, but the first
+	// datapoint is 1m).
+	InitialOffset int
+	// WindowSize sets the number of datapoints averaged out in a slope calculation.
```

Comment:
Yeah I guess technically they aren't averaged out, but if the values are monotonically non-decreasing, it's an equivalent calculation.

I'll add a comment clarifying that but keep the original - because I think "averaged out" is a much simpler thing to explain :sweat_smile:

Reply:
Added in 2f81286c1fa3eb7541eb738fbc3a754440245f91

Reply:
I don't understand why they are equivalent...

Code currently is `series[t]-series[t-w]`

"Averaged out" in my understanding would result in:
```
sum := 0
for( i := t-w; i < t; i++) {
  sum += series[i]
}
return sum/w
```

Are these equivalent?

Reply:
Ah I see what you mean. I was thinking from the perspective that the "slope" is averaged out (rather than the points), but I think your interpretation is probably more likely.

Reworded in c3b83a2aef62b275fda362d7051703f9e5640c54 & 7daeb3ec97d7dc6849c3c7614c7906c91883f2e8

> WindowSize sets the offset for datapoints used in the calculation of the slope before & after a point. For window size W, we calculate the slope at point P as value[P]-value[P-(W-1)].

Open to suggestions for further rewording.

Reply:
I think that's fine.

However, this `-1` is annoying. WDYT about this:

> SlopeDistance sets the distance between points used for slope calculation. At point i we would compare two slopes: (arr[i] - arr[i - SlopeDistance]) and (arr[i + SlopeDistance] - arr[i]). SlopeDistance must be >=1.

Reply:
discussed -- I'll explain the -1 in a comment next to it

Reply:
Added a comment in d1fb91592ef8dd88863038963e3d4b4499655f4c